main.py

```diff
@@ -29,6 +23,123 @@ def log_console_message(msg):
     print(f'Browser console: {msg}')
 
 
+def calculate_frame_durations_and_fps(*, trace_events: list):
```

Comment:
@maxrjones / @katamartin, 

i was able to extract frame durations and fps from the trace events.  it appears that the results are quite promising! I am curious to hear your thoughts on this.


<img width="1076" alt="Screenshot 2023-07-12 at 20 56 00" src="https://github.com/carbonplan/benchmark-maps/assets/13301940/f25bed62-5392-4924-957a-8c76c34b4e36">


<img width="1116" alt="Screenshot 2023-07-12 at 20 56 11" src="https://github.com/carbonplan/benchmark-maps/assets/13301940/f07115c1-ae9b-47f5-bd97-331751b57631">

<img width="1092" alt="Screenshot 2023-07-12 at 20 56 28" src="https://github.com/carbonplan/benchmark-maps/assets/13301940/86350c13-c23f-462e-a063-6046e3432d17">

<img width="1185" alt="Screenshot 2023-07-12 at 20 56 47" src="https://github.com/carbonplan/benchmark-maps/assets/13301940/1c9a95fd-5419-411b-895f-c1de25f42e03">

Reply:
This is really exciting, thank you @andersy005! I didn't get a chance to look in detail today, but tomorrow will take a look at how the frame durations recorded in data/data-{date}.json compare to the frame rates from chrome-devtools-traces/{date}.json as a QA/QC step.

Reply:
@andersy005 how did you determine that the frame durations are given by the difference in timesteps between consecutive 'Swap' events?

I'm also curious if you and @katamartin think the frame durations reported in the chrome dev tools performance viewer should be considered "truth" for these values? I ran the script locally and found that the results in `data-{date}.json` differ from those visualized in chrome dev tools from `chrome-devtools-traces/{date}.json `. For example, below is a histogram of the frame durations parsed by the script relative to those parsed by Chrome's performance viewer. While we care about the relative difference between runs more than the absolute values, I think this disparity is worth understanding to make sure the observed relative differences are meaningful.

![image](https://github.com/carbonplan/benchmark-maps/assets/14077947/5c3c04f3-6a1b-49da-b00e-b4d14c79f004)

Reply:
@maxrjones I talked about this a bit with @andersy005 last week. I mentioned that because we're deriving some frame durations < 16.67 ms something must be off in our calculation, potentially related to dropped frames. @andersy005 was going to test out different events (I think `BeginFrame` and `DroppedFrame` or something similar) and compare against the way Chrome parses the events to test for correctness. So in short, yes I think we should close or at least understand the gap to gain confidence in our results!